felt/src/lib.rs

```diff
@@ -343,31 +409,27 @@ impl Add<&Felt252> for u64 {
     type Output = Option<u64>;
 
     fn add(self, rhs: &Felt252) -> Option<u64> {
-        const PRIME_DIGITS_LE_HI: (u64, u64, u64) =
-            (0x0000000000000000, 0x0000000000000000, 0x0800000000000011);
-        const PRIME_MINUS_U64_MAX_DIGITS_LE_HI: (u64, u64, u64) =
-            (0xffffffffffffffff, 0xffffffffffffffff, 0x0800000000000010);
+        const PRIME_DIGITS_BE_HI: [u64; 3] =
+            [0x0800000000000011, 0x0000000000000000, 0x0000000000000000];
+        const PRIME_MINUS_U64_MAX_DIGITS_BE_HI: [u64; 3] =
+            [0x0800000000000010, 0xffffffffffffffff, 0xffffffffffffffff];
 
-        // Iterate through the 64 bits digits in little-endian order to
+        // Match with the 64 bits digits in big-endian order to
         // characterize how the sum will behave.
-        let mut rhs_digits = rhs.iter_u64_digits();
-        // No digits means `rhs` is `0`, so the sum is simply `self`.
-        let Some(low) = rhs_digits.next() else {
-            return Some(self);
-        };
-        // A single digit means this is effectively the sum of two `u64` numbers.
-        let Some(h0) = rhs_digits.next() else {
-            return self.checked_add(low)
-        };
-        // Now we need to compare the 3 most significant digits.
-        // There are two relevant cases from now on, either `rhs` behaves like a
-        // substraction of a `u64` or the result of the sum falls out of range.
-        let (h1, h2) = (rhs_digits.next()?, rhs_digits.next()?);
-        match (h0, h1, h2) {
+        match rhs.to_be_digits() {
+            // No digits means `rhs` is `0`, so the sum is simply `self`.
+            [0, 0, 0, 0] => Some(self),
+            // A single digit means this is effectively the sum of two `u64` numbers.
+            [0, 0, 0, low] => self.checked_add(low),
+            // Now we need to compare the 3 most significant digits.
+            // There are two relevant cases from now on, either `rhs` behaves like a
+            // substraction of a `u64` or the result of the sum falls out of range.
+
             // The 3 MSB only match the prime for Felt252::max_value(), which is -1
             // in the signed field, so this is equivalent to substracting 1 to `self`.
```

Comment:
The comments become misleading now :(
That description counting digits and all was quite specific to the iterative implementation. If doing digits directly it doesn't change that much.
In any case, we need to revalidate the premise that we need a specialized addition here with the new type. With `BigInt` was quite slow and a major bottleneck, but that's not necessarily still the case with `FieldElement`.